### PR TITLE
nixos/cryptpad: allow config with attrset

### PR DESCRIPTION
**Pull request by eadwu**
Compared generated `config.js` with the original `config.example.js`.
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/cryptpad.nix [hunk 1]
```diff
@@ -50,5 +76,23 @@ in
         WorkingDirectory = "%S/cryptpad";
       };
     };
+
+    services.cryptpad.settings = {
```
Comment (symphorien) [inline nixos/modules/services/web-apps/cryptpad.nix:80]: this should be the default value of the option, or at least be declared with `lib.mkDefault`

Comment (eadwu) [reply nixos/modules/services/web-apps/cryptpad.nix]: Should be fixed now.

Comment (symphorien) [reply nixos/modules/services/web-apps/cryptpad.nix]: I'll test this soon.
Is it really impossible to use the `default` key of mkOption ? This way the default value is shown on https://nixos.org/nixos/options.html.

**Comment by symphorien** [comment]
While reviewing this I got bit by the fact that
```
services.cryptpad = {
enable = true;
settings = { /* foo */ };
}
```
ignores the specified settings.

I wonder if we should do the following:
* make the default value of `configFile` null (the example config file can be an `example`)
* add an assertion that exactly one of `configFile` and `settings` is set to a non default value.


Otherwise I tested it and it works.

**Comment by symphorien** [comment]
I noticed while reviewing that I had the following warning:
```
juin 16 21:46:22 blitiri cryptpad[23293]: FRESH MODE ENABLED
juin 16 21:46:22 blitiri cryptpad[23293]:     m     m   mm   mmmmm  mm   m mmmmm  mm   m   mmm    m
juin 16 21:46:22 blitiri cryptpad[23293]:     #  #  #   ##   #   "# #"m  #   #    #"m  # m"   "   #
juin 16 21:46:22 blitiri cryptpad[23293]:     " #"# #  #  #  #mmmm" # #m #   #    # #m # #   mm   #
juin 16 21:46:22 blitiri cryptpad[23293]:      ## ##"  #mm#  #   "m #  # #   #    #  # # #    #
juin 16 21:46:22 blitiri cryptpad[23293]:      #   #  #    # #    " #   ## mm#mm  #   ##  "mmm"   #
juin 16 21:46:22 blitiri cryptpad[23293]: No 'httpSafeOrigin' provided.
juin 16 21:46:22 blitiri cryptpad[23293]: Your configuration probably isn't taking advantage of all of CryptPad's security features!
juin 16 21:46:22 blitiri cryptpad[23293]: This is acceptable for development, otherwise your users may be at risk.
juin 16 21:46:22 blitiri cryptpad[23293]: Serving sandboxed content via port 3001.
juin 16 21:46:22 blitiri cryptpad[23293]: This is probably not what you want for a production instance!
juin 16 21:46:22 blitiri cryptpad[23293]: Cryptpad is customizable, see customize.dist/readme.md for details
```

and yet I did provide 'httpSafeOrigin'.
i think this is fixed by https://github.com/xwiki-labs/cryptpad/commit/9b3013278bac5bb73be28cc4121bd56ab7a11e61
It looks like the warning is harmless.

It looks like cryptpad should be updated. (but not in this PR, I just mention this for potential reviewers).

**Comment by eadwu** [comment]
The only problem with using null by default is that the [new] defaults have to pulled in manually each release (if there are any new options).

**Comment by symphorien** [comment]
We would have to do this as well for the default value of the `settings` option...

**Comment by eadwu** [comment]
I see `settings` being used when the user probably knows exactly what options to set/enable. Ideally `settings` would pull in the new defaults if there were any, which should probably happen but not guaranteed, so using the bundled config file is probably still better.

**Comment by symphorien** [comment]
This distinction does not strike me as obvious, but ok.

**Comment by taku0** [comment]
How about making `settings` nullable rather than `configFile`, and use `settings` if it is not null.

**Comment by tomberek** [comment]
Functions as expected, but does require a significant nginx config get full function. See http://cgit.notk.org/asmadeus/nixos-config.git/tree/machines/jormungand

**Comment by tomberek** [comment]
@eadwu is there still interest in this? Or would want to use an RFC42-style approach?

**Comment by eadwu** [comment]
Well I can fix it up, but it probably doesn't matter since there doesn't seem to much interest anyway.

**Comment by Pamplemousse** [comment]
@eadwu Oh, thanks for starting this! I am very much interested!

**Comment by Pamplemousse** [comment]
@eadwu @tomberek
I have tried it out, and it works as expected (although setting up `nginx` is quite a pain).

I have made some minor adjustments, that could be integrated to this PR:
https://github.com/Pamplemousse/nixpkgs/commits/PR_89384

**Comment by McSinyx** [comment]
I am also interested, since this would allow more programmatic nginx configuration.

**Comment by drupol** [comment]
For some reason that I can't find, `cryptpad` has been removed from `nixpkgs`.

Therefore, closing this PR.

**Comment by McSinyx** [comment]
It is part of GH-180066 removing Node.js 12.

**Comment by drupol** [comment]
We could restore it if we want to, cryptpad is now using NodeJS 16... but NodeJS 16 is [EOL in 3 months](https://endoflife.date/nodejs). Maybe we should revisit this as soon as they switched to NodeJS 18.

**Comment by Pamplemousse** [comment]
> We could restore it if we want to, cryptpad is now using NodeJS 16

I remember that the main painpoint in packaging `cryptpad` was that they use `bower` to bundle their front-end dependencies. As far as I can remember, `bower` was introducing the dependency on `node` 12...

And so far, no hints on their part in getting rid of it: https://github.com/cryptpad/cryptpad/issues/295#issuecomment-1373267211 .

**Pull request closed without merging**